### PR TITLE
APIO-113 Add support for nested resources

### PR DESCRIPTION
**Pull request by nhpatt**

### CODE REVIEW

**Comment by nhpatt** [comment]
ci:test

File: apio-architect-message-hal/src/test/java/com/liferay/apio/architect/message/hal/internal/HALTestUtil.java [hunk 1]
```diff
@@ -224,8 +256,59 @@
...
+		).build();
+
+		Conditions conditions = builder.where(
+			"number1", is(aJsonInt(equalTo(42)))
+		).where(
+			"string1", is(aJsonString(equalTo("id 2")))
+		).where(
+			"_links", aJsonObjectWith(firstEmbeddedLinkConditions)
+		).where(
+			"_embedded", aJsonObjectWith(conditions2)
```
Comment (ahdezma) [inline apio-architect-message-hal/src/test/java/com/liferay/apio/architect/message/hal/internal/HALTestUtil.java]: you can use `aJsonObjectWhere`

File: apio-architect-message-hal/src/test/java/com/liferay/apio/architect/message/hal/internal/HALTestUtil.java [hunk 2]
```diff
@@ -224,8 +256,59 @@
...
+	 * @return the matcher
+	 * @review
+	 */
+	public static Matcher<JsonElement> isAJsonObjectWithTheSecondNested(
+		String id) {
+
+		Conditions.Builder builder = new Conditions.Builder();
+
+		Conditions conditions2 = builder.where(
+			"nested3", isAJsonObjectWithTheThirdNested()).build();
```
Comment (ahdezma) [inline apio-architect-message-hal/src/test/java/com/liferay/apio/architect/message/hal/internal/HALTestUtil.java]: you can use `aJsonObjectWhere`

File: apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java [hunk 3]
```diff
@@ -193,8 +217,50 @@
...
+		return is(aJsonObjectWith(conditions));
+	}
+
+	/**
+	 * Returns a {@code Matcher} that checks if the field is a JSON object of
+	 * the third nested model.
+	 *
+	 * @return the matcher
+	 */
+	public static Matcher<JsonElement> isAJsonObjectWithTheThirdNested() {
```
Comment (ahdezma) [inline apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java]: this method can be substituted by a call to `aJsonObjectWhere`

File: apio-architect-test-util/src/main/java/com/liferay/apio/architect/test/util/representor/MockRepresentorCreator.java [hunk 4]
```diff
@@ -157,6 +157,44 @@
 				"stringList1", __ -> asList("a", "b", "c", "d", "e")
 			).addStringList(
 				"stringList2", __ -> asList("f", "g", "h", "i", "j")
+			).addNestedField(
+				"nestedField1", rootModel -> (FirstEmbeddedModel)() -> "id 1",
```
Comment (ahdezma) [inline apio-architect-test-util/src/main/java/com/liferay/apio/architect/test/util/representor/MockRepresentorCreator.java]: you can substitute `rootModel` with `__`

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/PageWriter.java [hunk 5]
```diff
@@ -460,9 +464,19 @@ private void _writeItem(SingleModel<T> singleModel) {
...
 
+		_writeItemEmbeddedModelFields(
+			singleModel, embeddedPathElements, itemJsonObjectBuilder,
+			_representorFunction, singleModel);
+	}
+
+	private <S> void _writeItemEmbeddedModelFields(
+		SingleModel<S> singleModel, FunctionalList<String> embeddedPathElements,
+		JSONObjectBuilder itemJsonObjectBuilder,
+		RepresentorFunction representorFunction, SingleModel rootModel) {
```
Comment (ahdezma) [inline apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/PageWriter.java]: `rootModel` parameter should be `rootSingleModel`

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/PageWriter.java [hunk 6]
```diff
@@ -549,6 +563,43 @@ private void _writeItem(SingleModel<T> singleModel) {
...
 					resourceEmbeddedPathElements, url));
+
+		_writeNestedResources(
+			representorFunction, singleModel, itemJsonObjectBuilder, rootModel,
+			embeddedPathElements);
+	}
+
+	private <U> void _writeNestedResources(
+		RepresentorFunction representorFunction, SingleModel<U> singleModel,
+		JSONObjectBuilder itemJsonObjectBuilder, SingleModel rootModel,
```
Comment (ahdezma) [inline apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/PageWriter.java]: `rootModel` parameter should be `rootSingleModel`

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/PageWriter.java [hunk 7]
```diff
@@ -549,6 +563,43 @@ private void _writeItem(SingleModel<T> singleModel) {
...
+
+		representorOptional.ifPresent(
+			_representor -> {
+				Map<String, Representor<?, ?>> nestedFields =
+					_representor.getNestedFields();
+
+				nestedFields.forEach(
+					(key, value) -> {
+						Function<U, ?> nestedFieldMapper =
+							_representor.getNestedFieldFunctions().get(key);
```
Comment (ahdezma) [inline apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/PageWriter.java]: avoid chaining :(

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/PageWriter.java [hunk 8]
```diff
@@ -549,6 +563,43 @@ private void _writeItem(SingleModel<T> singleModel) {
...
+						Object mappedModel = nestedFieldMapper.apply(
+							singleModel.getModel());
+
+						FunctionalList embeddedNestedPathElements =
+							new FunctionalList(embeddedPathElements, key);
+
+						_writeItemEmbeddedModelFields(
+							new SingleModel<>(mappedModel, Object.class),
+							embeddedNestedPathElements, itemJsonObjectBuilder,
+							clazz -> Optional.of(value), rootModel);
```
Comment (ahdezma) [inline apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/PageWriter.java]: `clazz` can be renamed to `__`

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/SingleModelWriter.java [hunk 9]
```diff
@@ -489,6 +501,39 @@ public OperationsFunctionStep modelMessageMapper(
...
+
+		representorOptional.ifPresent(
+			_representor -> {
+				Map<String, Representor<?, ?>> nestedFields =
+					_representor.getNestedFields();
+
+				nestedFields.forEach(
+					(key, value) -> {
+						Function<S, ?> nestedFieldMapper =
+							_representor.getNestedFieldFunctions().get(key);
```
Comment (ahdezma) [inline apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/SingleModelWriter.java]: avoid chaining :(

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/SingleModelWriter.java [hunk 10]
```diff
@@ -489,6 +501,39 @@ public OperationsFunctionStep modelMessageMapper(
...
+						Object mappedModel = nestedFieldMapper.apply(
+							singleModel.getModel());
+
+						FunctionalList<String> embeddedNestedPathElements =
+							new FunctionalList<>(embeddedPathElements, key);
+
+						writeEmbeddedModelFields(
+							new SingleModel<>(mappedModel, Object.class),
+							embeddedNestedPathElements,
+							clazz -> Optional.of(value));
```
Comment (ahdezma) [inline apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/SingleModelWriter.java]: `clazz` can be renamed to `__`

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java [hunk 11]
```diff
@@ -81,13 +106,59 @@
...
+
 		Optional<Representor<T, ?>> optional = getRepresentorOptional(
 			singleModel.getModelClass(), representorFunction);
 
 		return optional.flatMap(
-			representor -> pathFunction.apply(
-				representor.getIdentifier(singleModel.getModel()),
-				representor.getIdentifierClass(), singleModel.getModelClass()));
+			representor -> {
+				if (representor.getIdentifierFunction() == null) {
```
Comment (ahdezma) [inline apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java]: can this be simplified with a chaining of `filter`-`map`-`orElseGet`?

File: apio-architect-api/src/main/java/com/liferay/apio/architect/representor/Representor.java [hunk 12]
```diff
@@ -221,6 +259,21 @@ public Builder(
 				identifierClass, relatedCollectionsSupplier);
 		}
 
+		/**
+		 * Adds a type for a nested resource, skips identifier function because
+		 * nested resources does not have an ID
+		 *
+		 * @param  type the type name
+		 * @return the builder's step
```
Comment (ahdezma) [inline apio-architect-api/src/main/java/com/liferay/apio/architect/representor/Representor.java]: `@review`?

File: apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java [hunk 13]
```diff
@@ -154,8 +158,28 @@
 	}
 
 	/**
-	 * Returns a {@code Matcher} that checks if the field is a JSON
-	 * object of the second embedded model.
+	 * Returns a {@code Matcher} that checks if the field is a JSON object of
+	 * the first nested model.
```
Comment (ahdezma) [inline apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java]: `@review`?

File: apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java [hunk 14]
```diff
@@ -193,8 +217,50 @@
 	}
 
 	/**
-	 * Returns a {@code Matcher} that checks if the field is a link
-	 * to the URL.
+	 * Returns a {@code Matcher} that checks if the field is a JSON object of
+	 * the second nested model.
+	 *
+	 * @return the matcher
```
Comment (ahdezma) [inline apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java]: `@review`?

File: apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java [hunk 15]
```diff
@@ -193,8 +217,50 @@
...
+		).build();
+
+		return is(aJsonObjectWith(conditions));
+	}
+
+	/**
+	 * Returns a {@code Matcher} that checks if the field is a JSON object of
+	 * the third nested model.
+	 *
+	 * @return the matcher
```
Comment (ahdezma) [inline apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java]: `@review`?

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java [hunk 16]
```diff
@@ -51,12 +51,37 @@
...
+	 * Returns the {@link FieldsWriter} for a given model.
+	 *
+	 * @param  singleModel the single model
+	 * @param  embeddedPathElements the embedded path element list
+	 * @param  requestInfo the current request's information
+	 * @param  pathFunction the function to get the {@link Path}
+	 * @param  rootRepresentorFunction the function to get the {@link
+	 *         Representor} of the parent model
+	 * @param  rootSingleModel the parent model
+	 * @return the {@code FieldsWriter} for the model
```
Comment (ahdezma) [inline apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java]: `@review`?

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java [hunk 17]
```diff
@@ -81,13 +106,59 @@
...
+	 * Optional#empty()}.
+	 *
+	 * @param  singleModel the single model
+	 * @param  pathFunction the function that gets the {@code Path}
+	 * @param  representorFunction the function that gets the {@code
+	 *         Representor}
+	 * @param  rootRepresentorFunction the function that gets the {@code
+	 *         Representor} for the parent model
+	 * @param  rootSingleModel the parent model
+	 * @return the model's {@code Path}, if the model's {@code Representor} and
```
Comment (ahdezma) [inline apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java]: `@review`?

File: apio-architect-api/src/main/java/com/liferay/apio/architect/representor/Representor.java [hunk 18]
```diff
@@ -100,6 +101,16 @@ public S getIdentifier(T model) {
 		return _identifierClass;
 	}
 
+	/**
+	 * Returns the model's identifier function, the method that extracts the
+	 * identifier from the model
+	 *
+	 * @return the model's identifier function
```
[the remark this answers is not in the document]
Comment (nhpatt) [reply apio-architect-api/src/main/java/com/liferay/apio/architect/representor/Representor.java]: https://github.com/liferay/com-liferay-apio-architect/pull/145/commits/4d822f69ac21287de3eb007615222b6063ca68fc

File: apio-architect-api/src/main/java/com/liferay/apio/architect/representor/Representor.java (hunk 12 above)
Comment (nhpatt) [reply apio-architect-api/src/main/java/com/liferay/apio/architect/representor/Representor.java]: https://github.com/liferay/com-liferay-apio-architect/pull/145/commits/4d822f69ac21287de3eb007615222b6063ca68fc

File: apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java (hunk 13 above)
Comment (nhpatt) [reply apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java]: https://github.com/liferay/com-liferay-apio-architect/pull/145/commits/4d822f69ac21287de3eb007615222b6063ca68fc

File: apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java (hunk 14 above)
Comment (nhpatt) [reply apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java]: https://github.com/liferay/com-liferay-apio-architect/pull/145/commits/4d822f69ac21287de3eb007615222b6063ca68fc

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java (hunk 16 above)
Comment (nhpatt) [reply apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java]: https://github.com/liferay/com-liferay-apio-architect/pull/145/commits/4d822f69ac21287de3eb007615222b6063ca68fc

File: apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java (hunk 17 above)
Comment (nhpatt) [reply apio-architect-writer-api/src/main/java/com/liferay/apio/architect/writer/util/WriterUtil.java]: https://github.com/liferay/com-liferay-apio-architect/pull/145/commits/4d822f69ac21287de3eb007615222b6063ca68fc

File: apio-architect-message-hal/src/test/java/com/liferay/apio/architect/message/hal/internal/HALTestUtil.java (hunk 2 above)
Comment (nhpatt) [reply apio-architect-message-hal/src/test/java/com/liferay/apio/architect/message/hal/internal/HALTestUtil.java]: https://github.com/liferay/com-liferay-apio-architect/pull/145/commits/9fa98d0db1308408ebc9fa51df32320cec506dde

File: apio-architect-message-hal/src/test/java/com/liferay/apio/architect/message/hal/internal/HALTestUtil.java [hunk 19]
```diff
@@ -137,15 +140,44 @@
...
-				"embedded1", is(aJsonObjectWith(firstEmbeddedConditions))));
+		Conditions conditions = builder.where(
+			"embedded1", is(aJsonObjectWith(firstEmbeddedConditions))
+		).where(
+			"nestedField1", isAJsonObjectWithTheFirstNested()
+		).where(
+			"nestedField2", isAJsonObjectWithTheSecondNested(id)
+		).build();
+
+		Matcher<JsonElement> embedded1 = is(aJsonObjectWith(conditions));
```
[the remark this answers is not in the document]
Comment (nhpatt) [reply apio-architect-message-hal/src/test/java/com/liferay/apio/architect/message/hal/internal/HALTestUtil.java]: https://github.com/liferay/com-liferay-apio-architect/pull/145/commits/9fa98d0db1308408ebc9fa51df32320cec506dde

File: apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java (hunk 15 above)
Comment (nhpatt) [reply apio-architect-message-json-plain/src/test/java/com/liferay/apio/architect/message/json/plain/internal/PlainJSONTestUtil.java]: https://github.com/liferay/com-liferay-apio-architect/commit/4d822f69ac21287de3eb007615222b6063ca68fc

**Comment by nhpatt** [comment]
ci:test

**Comment by nhpatt** [comment]
ci:stop

**Comment by nhpatt** [comment]
ci:test

**Comment by nhpatt** [comment]
ci:test

**Pull request merged**